File: tests/39-basic-api_level.py
#!/usr/bin/env python

#
# Seccomp Library test program
#
# Copyright (c) 2016 Red Hat <pmoore@redhat.com>
# Copyright (c) 2017 Canonical Ltd.
# Authors: Paul Moore <paul@paul-moore.com>
#          Tyler Hicks <tyhicks@canonical.com>
#

#
# This library is free software; you can redistribute it and/or modify it
# under the terms of version 2.1 of the GNU Lesser General Public License as
# published by the Free Software Foundation.
#
# This library is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
# for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library; if not, see <http://www.gnu.org/licenses>.
#

import argparse
import sys

import util

from seccomp import *

def test():
    api = get_api()
    if (api < 1):
        raise RuntimeError("Failed getting initial API level")

    set_api(1)
    api = get_api()
    if api != 1:
        raise RuntimeError("Failed getting API level 1")

    set_api(2)
    api = get_api()
    if api != 2:
        raise RuntimeError("Failed getting API level 2")

    set_api(3)
    api = get_api()
    if api != 3:
        raise RuntimeError("Failed getting API level 3")

    set_api(4)
    api = get_api()
    if api != 4:
        raise RuntimeError("Failed getting API level 4")

    set_api(5)
    api = get_api()
    if api != 5:
        raise RuntimeError("Failed getting API level 5")

    set_api(6)
    api = get_api()
    if api != 6:
        raise RuntimeError("Failed getting API level 6")

    # Attempt to set a high, invalid API level
    try:
        set_api(1024)
    except ValueError:
        pass
    else:
        raise RuntimeError("Missing failure when setting invalid API level")
    # Ensure that the previously set API level didn't change
    api = get_api()
    if api != 6:
        raise RuntimeError("Failed getting old API level after setting an invalid API level")

test()

# kate: syntax python;
# kate: indent-mode python; space-indent on; indent-width 4; mixedindent off;

File: tests/39-basic-api_level.c
/**
 * Seccomp Library test program
 *
 * Copyright (c) 2017 Red Hat <pmoore@redhat.com>
 * Author: Paul Moore <paul@paul-moore.com>
 */

/*
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of version 2.1 of the GNU Lesser General Public License as
 * published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, see <http://www.gnu.org/licenses>.
 */

#include <errno.h>
#include <unistd.h>

#include <seccomp.h>

int main(int argc, char *argv[])
{
	int rc;
	unsigned int api;

	api = seccomp_api_get();
	if (api < 1)
		return -1;

	rc = seccomp_api_set(1);
	if (rc != 0)
		return -2;
	api = seccomp_api_get();
	if (api != 1)
		return -3;

	rc = seccomp_api_set(2);
	if (rc != 0)
		return -4;
	api = seccomp_api_get();
	if (api != 2)
		return -5;

	rc = seccomp_api_set(3);
	if (rc != 0)
		return -6;
	api = seccomp_api_get();
	if (api != 3)
		return -7;

	rc = seccomp_api_set(4);
	if (rc != 0)
		return -8;
	api = seccomp_api_get();
	if (api != 4)
		return -9;

	rc = seccomp_api_set(5);
	if (rc != 0)
		return -10;
	api = seccomp_api_get();
	if (api != 5)
		return -11;

	rc = seccomp_api_set(6);
	if (rc != 0)
		return -12;
	api = seccomp_api_get();
	if (api != 6)
		return -13;

	/* Attempt to set a high, invalid API level */
	rc = seccomp_api_set(1024);
	if (rc != -EINVAL)
		return -1001;
	/* Ensure that the previously set API level didn't change */
	api = seccomp_api_get();
	if (api != 6)
		return -1002;

	return 0;
}
